src/shape.cpp

```diff
@@ -49,8 +49,13 @@ Dimension::Dimension(dimension_value_t dim) :
 }
 
 Dimension::Dimension(dimension_value_t minimum, dimension_value_t maximum) {
-    // TODO we should either throw or make separate factory to disallo
-    // calling with eg (10,9) or (-2,-1) etc
+    if (minimum < -1 || maximum < -1) {
```

Comment:
Add unit tests

Reply:
What should be expected behavior for:
Dimension(-1,0)
and
Dimension(0,0)?
Neither throws. I think the firs one should (since original intention is to -1 encode "totally dynamic" dimension. With (0,0) I am not sure since on one hand it would denote empty tensor. I think it should be accepted so we could use it eg. in empty tensor in DAG in the future.
@dkalinowski  what's your take?